### ISSUE TITLE
Data Loss Test for Pause Feature

### ISSUE THREAD

**Issue by juliev0** [numaproj/numaplane#140]
# Summary

We need to gain confidence that the Pause Feature is doing what we expect. Ideally, we'd design a test that could cause data loss, but I'm not sure how difficult that would be. Alternatively, we may just need to verify that pausing happened when it was supposed to. 

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by juliev0** [comment]
I am moving the "Load Test" that was in the original description out of here and making this issue strictly related to a test for "data loss".

**Comment by juliev0** [comment]
I have just now performed a basic test of a Pipeline update in which a vertex gets removed. In this case, I use a vertex which intentionally sleeps, so it builds up a lot of data in its buffer. 

If there were no "data loss prevention", the data would get lost. But in this case, the data is drained through all of the way and only once the Numaflow Controller has finished and declared the Pipeline as "Paused" do we resume the Pipeline Running. 

I'm attaching the artifacts here:
[Archive.zip](https://github.com/user-attachments/files/16790042/Archive.zip)

The source Pod is an http listener. We send an auto-incrementing integer through a bash script to it. Upon failure, the bash script continues to retry the same integer. The bash script sends at a slightly higher rate than the "cat" Vertex in the Pod sleeps, to build up a buffer. We want to confirm that the Pipeline drains all messages that were successfully sent to the Source Pod before the Pipeline goes back to "Running" state. 

Instructions:

1. `kubectl apply -f PipelineRollout1.yaml`
2. Port forward "in" (source) Pod (port 8443)
3. Tail "out" (sink) log
4. `bash curl-auto-increment.sh`
5. See a new line printed with the next integer in the "out" log every 10 seconds. After a few seconds, see that there are a few more messages that have been sent to Source Pod (you can tell by reading the output from the bash script) than the "out" log. 
6. kubectl apply -f PipelineRollout2.yaml
7. Watch the Pipeline pause while the "out" log gradually prints the remaining messages
8. Watch the Pipeline enter Paused and then immediately re-enter Running state once all messages have been printed

**Comment by juliev0** [comment]
I have now performed a test in which 2 Pipelines are running, which are just like the Pipeline from the previous test, where one vertex is a very slow "cat" vertex (it sleeps), but in this case I performed a Numaflow Controller version update and observed that both of them paused correctly. One bash script curled to one Pipeline, and another bash script to the other. 

Once I tried to update the Numaflow Controller, both pipelines paused until all of the data that was sent to them went through, and only then did the Numaflow Controller Pod update, and then both pipelines resumed. I observed that the new Numaflow Controller Pod's creation timestamp preceded the timestamp of all of the pipeline Pods.

I then repeated the same test but for updating the ISBService.
